### PR TITLE
Adding support for XDG Base Directory specification

### PR DESCRIPTION
**Pull request by azatoth**
Enable the program to follow the basedir spec per
http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

### CODE REVIEW

File: libobs/util/platform-nix.c [hunk 1]
```diff
@@ -298,12 +434,36 @@ int os_unlink(const char *path)
 
 int os_mkdir(const char *path)
 {
-	if (mkdir(path, 0777) == 0)
+	if (mkdir(path, 0755) == 0)
 		return MKDIR_SUCCESS;
-
```
Comment (palana) [inline libobs/util/platform-nix.c]: Please don't delete (empty) lines randomly

File: libobs/util/platform.h [hunk 2]
```diff
@@ -92,6 +92,10 @@ EXPORT uint64_t os_gettime_ns(void);
 
 EXPORT int os_get_config_path(char *dst, size_t size, const char *name);
 EXPORT char *os_get_config_path_ptr(const char *name);
+EXPORT int os_get_data_path(char *dst, size_t size, const char *name);
```
Comment (palana) [inline libobs/util/platform.h]: The addition of new API calls should happen in a separate commit from adding support for a new library

File: libobs/util/platform.h [hunk 3]
```diff
@@ -132,6 +136,8 @@ EXPORT int os_unlink(const char *path);
 
 EXPORT int os_mkdir(const char *path);
 
+EXPORT int os_mkdirs(const char *path, int mode);
```
Comment (palana) [inline libobs/util/platform.h]: This should be in a separate commit.

File: libobs/util/platform-cocoa.m [hunk 4]
```diff
@@ -81,6 +81,14 @@ int os_get_config_path(char *dst, size_t size, const char *name)
 	return snprintf(dst, size, "%s/%s", base_path, name);
 }
 
+int os_get_data_path(char *dst, size_t size, const char *name) {
```
Comment (palana) [inline libobs/util/platform-cocoa.m]: All new functions seem to violate the coding style: opening braces for functions belong on a new line, e.g.:
```cpp
int os_get_data_path(char *dst, size_t size, const char *name)
{
```

File: libobs/util/platform-nix.c [hunk 5]
```diff
@@ -153,27 +153,163 @@ uint64_t os_gettime_ns(void)
 	return ((uint64_t) ts.tv_sec * 1000000000ULL + (uint64_t) ts.tv_nsec);
 }
 
-/* should return $HOME/.[name] */
+/* should return $HOME/.[name], or when using XDG, should return $HOME/.config/[name] as default */
```
Comment (palana) [inline libobs/util/platform-nix.c]: Please keep all lines below 80 columns

File: libobs/util/platform-nix.c [hunk 6]
```diff
@@ -298,12 +457,37 @@ int os_unlink(const char *path)
...
 int os_mkdir(const char *path)
 {
-	if (mkdir(path, 0777) == 0)
+	if (mkdir(path, 0755) == 0)
 		return MKDIR_SUCCESS;
 
 	return (errno == EEXIST) ? MKDIR_EXISTS : MKDIR_ERROR;
 }
 
+int os_mkdirs(const char *path, int mode) {
```
Comment (jp9000) [inline libobs/util/platform-nix.c]: This wouldn't port to windows.  You cannot set the permissions like this.  It's why os_mkdir does not have this parameter in the first place.

Comment (jp9000) [inline libobs/util/platform-nix.c]: Ah okay I see that you're just ignoring it on windows.  Apologies about that.

**Comment by jp9000** [comment]
Going to look at this more in depth at some point, just need to make sure old settings can be preserved

**Comment by kc5nra** [comment]
ok to test

**Comment by ONSBalder** [comment]
Can one of the admins verify this patch?

**Comment by Streusel** [comment]
@jp9000 close? pull? status?

**Comment by jp9000** [comment]
I left it open because there were changes that had to be made (due to the fact that it doesn't preserve any of your settings), but I was interested in eventually getting it merged.  I was really busy for the last patch, and now I will have free time again to see if I can get it merged (depending on how it functions).
